src/runtime/loader.ts

```diff
@@ -0,0 +1,98 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+import * as url from 'url';
+import * as path from 'path';
+
+import { ManifestBackend, ManifestEndpoint } from './manifest';
+
+/**
+ * Dynamically load import function to prevent TypeScript from
+ * transpiling into a require.
+ *
+ * See https://github.com/microsoft/TypeScript/issues/43329.
+ */
+const dynamicImport = new Function(
+  'modulePath',
+  'return import(modulePath)'
+) as (modulePath: string) => Promise<any>;
+
+async function loadModule(functionsDir: string) {
+  const absolutePath = path.resolve(functionsDir);
+  try {
+    return require(path.resolve(absolutePath));
+  } catch (e) {
+    if (e.code === 'ERR_REQUIRE_ESM') {
+      // This is an ESM package!
+      const modulePath = require.resolve(absolutePath);
+      // Resolve module path to file:// URL. Required for windows support.
+      // @ts-ignore pathToFileURL exists for Node.js v10 and up. Since ESM support exists for Node.js v13 and up, we
+      // can be sure that this function exists here.
+      const moduleURL = url.pathToFileURL(modulePath).href;
+      return await dynamicImport(moduleURL);
+    }
+    throw e;
+  }
+}
+
+/* @internal */
+export function extractEndpoints(
+  module,
+  endpoints: Record<string, ManifestEndpoint>,
+  prefix = ''
+) {
+  for (const [name, val] of Object.entries(module)) {
+    if (
+      typeof val === 'function' &&
+      val['__endpoint'] &&
+      typeof val['__endpoint'] === 'object'
+    ) {
+      const funcName = prefix + name;
+      endpoints[funcName] = {
+        ...val['__endpoint'],
+        entryPoint: funcName.replace(/-/g, '.'),
+      };
+    } else if (typeof val === 'object' && val !== null) {
+      extractEndpoints(val, endpoints, prefix + name + '-');
+    }
+  }
+}
+
+/* @internal */
+export async function loadBackend(
+  functionsDir: string
+): Promise<ManifestBackend> {
+  const endpoints: Record<string, ManifestEndpoint> = {};
+  const mod = await loadModule(functionsDir);
+
+  extractEndpoints(mod, endpoints);
+
+  const backend: ManifestBackend = {
+    endpoints,
+    specVersion: 'v1alpha1',
+    requiredAPIs: {},
+  };
+  if (Object.values(endpoints).find((ep) => ep.scheduleTrigger)) {
+    backend.requiredAPIs['pubsub'] = 'pubsub.googleapis.com';
```

Comment:
I had map <friendly name, API> in the CLI because I thought we were supposed to print "pubsub" on deploy. I've since learned that we should print "functions" instead, so the key here is useless.

OTOH, I thought we _also_ wanted to add "reason" here. So the manifest would be

```yaml
requiredAPIs:
  - api: pubsub.googleapis.com
     reason: needed for v1 scheduled functions
  - api: cloudscheduler.googleapis.com
    reason: needed for scheduled functions
```

Reply:
Done!